impute_constant/impute_constant.tex

```diff
@@ -0,0 +1,181 @@
+\begin{enumerate}
+    \item \textup{(Appropriate output domain).} If vector $v$ is in the \texttt{input\_domain}, then \texttt{function(v)} is in the \texttt{output\_domain}.
+    \item \textup{(Domain-Metric Compatibility).} The domain \texttt{input\_domain} matches one of the possible domains listed in the definition of \texttt{input\_metric}, and likewise \texttt{output\_domain} matches one of the possible domains listed in the definition of \texttt{output\_metric}.
+    \item \textup{(Stability Guarantee).} For every pair of elements $v, w$ in \texttt{input\_domain} and for every pair $(\din, \dout)$, where $\din$ is of the associated type for \texttt{input\_metric} and $\dout$ is the associated type for \texttt{output\_metric}, if $v,w$ are $d_{in}$-close under \texttt{input\_metric} and $\Relation(\din, \dout) = \True$, then $\function(v), \function(w)$ are $d_{out}$-close under \texttt{output\_metric}.
+\end{enumerate}
+\end{theorem}
+\begin{proof}
+\begin{enumerate}
+
+\item \textbf{(Appropriate output domain).} 
+\grace{Things to confirm: (1) is the type signature correct (input and output type)? The input and output domain correct? (2) What is the reason the type signature is not sufficient? Is it because the output type is more general than the output domain? Or is it theoretically because constant can be null value even if it has the type DA::NonNull? (3) so, is it correct that a variable with type DA::NonNull can contain null values?}
+
+
+In the case of \texttt{make\_impute\_constant}, this corresponds to showing that for every vector $v$ of elements of type \texttt{DA},  \texttt{function}$(v)$ is a vector of elements of type \texttt{DA::NonNull}. We can also say that \texttt{function}$(v)$ is a vector of elements that does not contain any \texttt{NonNull} values.
+
+The $\function(v)$ has type \texttt{Vec(DA)} follows from the assumption that element $v$ is in \texttt{input\_domain} and from the type signature of \texttt{function} in line~\ref{line:fn} of the pseudocode (Section~\ref{sec:pseudocode}), which takes in an element of type \texttt{Vec(DA)} and returns an element of type \texttt{Vec(DA::NonNull)}. If the Rust code compiles correctly, then the type correctness follows from the definition of the type signature enforced by Rust. Otherwise, the code raises an exception for incorrect input type. 
+
+Lemma \ref{lemma-NonNull} tells us the type signature is not sufficient check because (1) the output type of \texttt{VectorDomain(DA::NonNull)} is more general than the output domain of \texttt{VectorDomain(DA::NonNull)}. \grace{?? The output type matches the output domain I wrote, so something's wrong.} (2) \texttt{constant} can be \texttt{null} even if it is of type \texttt{DA::NonNull}. For (2), if the constant is null, then \texttt{impute\_constant} function (line \ref{line:atomfn}) could potentially return \texttt{null} values since it replaces a \texttt{null} input with \texttt{constant}.
```

Comment:
2. for appropriate output domain: 

What is the reason the type signature is not sufficient? Is it because the output type is more general than the output domain? Or is it theoretically because constant can be null value even if it has the type DA::NonNull?

Reply:
Hmm, `Vec(DA)` or "a vector of atomic domains" doesn't really make sense in this context. I think you meant `Vec(DA::Carrier)` in these blocks instead of `Vec(DA)`. 

And yes, in this case the output type can represent a value (float nan) that is not a member of the output domain. The purpose of the impute function is to ensure that these outputs are never null at runtime, so it replaces null with a constant. Since `constant` has type DA::NonNull, and DA::NonNull may itself by f64, an additional check is necessary that `constant` is non-null. We wouldn't want to allow analysts to impute nulls by replacing with nulls, violating our domains.

Reply:
Thank you Mike. This helps clarify things a lot!